src/slurm_plugin/common.py

```diff
@@ -8,5 +8,183 @@
+
+        instances_to_launch = self._parse_requested_instances()
+        for queue, queue_instances in instances_to_launch.items():
+            for instance_type, slurm_node_list in queue_instances.items():
+                log.info("Launching instances for slurm nodes %s", slurm_node_list)
+                for batch_nodes in grouper(slurm_node_list, self.max_batch_size):
+                    try:
+                        launched_instances = InstanceLauncher._launch_ec2_instances(
+                            ec2_client, self.cluster_name, queue, instance_type, len(batch_nodes)
+                        )
+                        if self.update_node_address:
+                            instance_ids, instance_ips, instance_hostnames = InstanceLauncher._parse_launched_instances(
+                                launched_instances
+                            )
+                            self._update_slurm_node_addrs(
+                                list(batch_nodes), instance_ids, instance_ips, instance_hostnames
+                            )
+                    except Exception as e:
+                        log.error(
+                            "Encountered exception when launching instances for nodes %s: %s", list(batch_nodes), e
+                        )
+                        self.failed_nodes.extend(batch_nodes)
+
+    def _update_slurm_node_addrs(self, slurm_nodes, instance_ids, instance_ips, instance_hostnames):
+        """Update node information in slurm with info from launched EC2 instance."""
+        try:
+            update_nodes(slurm_nodes, nodeaddrs=instance_ips, nodehostnames=instance_hostnames, raise_on_error=True)
+            log.info(
+                "Nodes %s are now configured with instance=%s private_ip=%s nodehostname=%s",
+                slurm_nodes,
+                instance_ids,
+                instance_ips,
+                instance_hostnames,
+            )
+        except subprocess.CalledProcessError:
+            log.error(
+                "Encountered error when updating node %s with instance=%s private_ip=%s nodehostname=%s",
+                slurm_nodes,
+                instance_ids,
+                instance_ips,
+                instance_hostnames,
+            )
+            self.failed_nodes.extend(slurm_nodes)
+
+    @staticmethod
+    def _parse_launched_instances(launched_instances):
+        """Parse run_instance output."""
+        instance_ids = []
+        instance_ips = []
+        instance_hostnames = []
+        for instance in launched_instances:
+            instance_ids.append(instance["InstanceId"])
+            instance_ips.append(instance["PrivateIpAddress"])
+            instance_hostnames.append(instance["PrivateDnsName"].split(".")[0])
+        return instance_ids, instance_ips, instance_hostnames
+
+    def _parse_requested_instances(self):
+        """
+        Parse out instance type from slurm NodeName.
```

Comment:
is this method really doing this? It is more counting how many instances of a certain type for each queue are requested.

Reply:
Done